research/picai/nnunet_scripts/eval.py

```diff
@@ -1,26 +1,188 @@
+        for cls in range(1, num_classes):  # Ignore background class at idx 0
+            cls_probs.append(extract_lesion_candidates(prob_map[cls], num_lesions_to_extract=5)[0])
+        detection_maps.append(np.stack(cls_probs))
+
+    return np.stack(detection_maps)
+
+
+def get_picai_metrics(
+    detection_maps: NDArray,
+    ground_truth_annotations: NDArray,
+    case_identifiers: Optional[Iterable[str]] = None,
+    **kwargs: Any
+) -> PicaiEvalMetrics:
+    """
+    Computes the picai evaluation metrics provided the predicted lesion
+    detection maps and the ground truth annotations. Extends picai_eval to
+    allow multiclass evaluation
+
+    Args:
+        detection_maps (NDArray): The predicted lesion detection maps. Must
+            have shape (num_samples, num_lesion_classes, ...). The background
+            class should not be included in the detection maps
+        ground_truth_annotations (NDArray): The ground truth annotations. Must
+            have shape (num_samples, num_classes or num_lesion_classes, ...).
+            If num_classes is provided, the function will attempt to remove
+            the background class from index 0 for you
+        case_identifiers (Optional[Iterable[str]], optional): A list of case
+            identifiers. If not provided the subjects will be identified by
+            their index Defaults to None.
+        **kwargs: Keyword arguments for the picai_eval.evaluate function
+
+    Raises:
+        KeyError: If you try to use the y_det, y_true, or subject_list keyword
+            arguments
+
+    Returns:
+        picai_eval.metrics.Metrics: A picai eval metrics object that has
+            combined the results from all classes into a single object
+    """
+    if "y_det" in kwargs or "y_true" in kwargs or "subject_list" in kwargs:
+        raise KeyError(
+            """Got one of 'y_det', 'y_true' or 'subject_list' in keyword
+            arguments. The arguments to get_picai_metrics are being passed to
+            these keys"""
+        )
+
+    if detection_maps.shape[1] != ground_truth_annotations.shape[1]:
+        # Assume background class has not been removed
+        ground_truth_annotations = ground_truth_annotations[:, 1:]
+        assert (
+            detection_maps.shape == ground_truth_annotations.shape
+        ), "Got unexpected shapes for detection maps and ground truth annotations"
+
+    # Evaluation must be calculated seperately for each class
+    num_classes = detection_maps.shape[1]
+    metrics: List[PicaiEvalMetrics] = []
+    for cls in range(num_classes):
+        metrics.append(
+            evaluate(
+                y_det=detection_maps[:, cls],
+                y_true=ground_truth_annotations[:, cls],
+                subject_list=case_identifiers,
+                **kwargs
+            )
+        )
+
+    subject_list = metrics[0].subject_list
+    assert isinstance(subject_list, list), "Got unexpected subject list from picai eval metrics object"
```

Comment:
Just curious for my understanding, in what cases would this subject_list not be a list?

Reply:
Its defined as optional in the metrics class, if none they generate one automatically. Mypy would complain so I had to narrow the type

Reply:
this code is no longer in the new script